src/Features/Core/Portable/UseNameOf/AbstractUseNameOfDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,101 @@
+﻿namespace Microsoft.CodeAnalysis.UseNameof
+{
+    using System.Collections.Immutable;
+    using System.Linq;
+    using System.Threading;
+    using Microsoft.CodeAnalysis.CodeStyle;
+    using Microsoft.CodeAnalysis.Diagnostics;
+    using Microsoft.CodeAnalysis.Editing;
+    using Microsoft.CodeAnalysis.LanguageServices;
+    using Microsoft.CodeAnalysis.Operations;
+
+    internal abstract class AbstractUseNameOfDiagnosticAnalyzer : AbstractCodeStyleDiagnosticAnalyzer
+    {
+        protected AbstractUseNameOfDiagnosticAnalyzer()
+            : base(
+                IDEDiagnosticIds.UseNameofDiagnosticId,
+                "Use nameof.",
+                "Use nameof.",
+                false)
+        {
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory() => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        public override bool OpenFileOnly(Workspace workspace) => false;
+
+        protected abstract ISyntaxFactsService GetSyntaxFactsService();
+
+        protected override void InitializeWorker(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.RegisterOperationAction(Handle, OperationKind.Literal);
+        }
+
+        private void Handle(OperationAnalysisContext context)
+        {
+            if (context.Operation is ILiteralOperation literal &&
+                literal.Parent is IArgumentOperation &&
+                literal.Type.SpecialType == SpecialType.System_String &&
+                literal.ConstantValue.HasValue &&
+                literal.ConstantValue.Value is string value &&
+                GetSyntaxFactsService() is ISyntaxFactsService syntaxFacts &&
```

Comment:
please don't use this 'pattern' (no pun intended) to just introduce a variable inside a large expression.  It's fine to literally just do ```var syntaxFacts = GetSyntaxFactsService()``` either before you start checking, or once you need it.

Reply:
We use it in the row below but changing it to two calls.